Fix failing h2spec tests 8.1.2.1 related to pseudo-headers validation

madgnome opened the pull request:
**Motivation:**

According to the [spec](http://httpwg.org/specs/rfc7540.html#HttpHeaders):

> All pseudo-header fields MUST appear in the header block before regular
> header fields. Any request or response that contains a pseudo-header
> field that appears in a header block after
> a regular header field MUST be treated as malformed (Section 8.1.2.6).

> Pseudo-header fields are only valid in the context in which they are defined.
> Pseudo-header fields defined for requests MUST NOT appear in responses;
> pseudo-header fields defined for responses MUST NOT appear in requests.
> Pseudo-header fields MUST NOT appear in trailers.
> Endpoints MUST treat a request or response that contains undefined or
> invalid pseudo-header fields as malformed (Section 8.1.2.6).

> Clients MUST NOT accept a malformed response. Note that these requirements
> are intended to protect against several types of common attacks against HTTP;
> they are deliberately strict because being permissive can expose
> implementations to these vulnerabilities.

**Modifications:**

- Introduce new constructor to DefaultHttp2Headers allowing
  header insertion order to be preserved (useful for testing)
- Assign a PseudoHeaderType (REQUEST|RESPONSE) to PseudoHeaderName
- Group variables related to header decoding to the inner class
  HeadersDecodingState
- Add pseudo-header validation to DefaultHttp2HeaderDecoder

**Result:**

- Requests with unknown pseudo-field headers are rejected
- Requests with containing response specific pseudo-headers are rejected
- Requests where pseudo-header appear after regular header are rejected
- Fixes #5761 tests 8.1.2.1

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Headers.java:
```diff
@@ -70,6 +70,7 @@ public void validateName(CharSequence name) {
     };
 
     private HeaderEntry<CharSequence, CharSequence> firstNonPseudo = head;
+    private boolean keepInsertionOrder;
```
final ?

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:
```diff
@@ -498,4 +477,46 @@ static long decodeULE128(ByteBuf in, long result) throws Http2Exception {
+        private final Http2Headers headers;
+        private final int streamId;
+
+        private HeadersDecodingState(final int streamId, final Http2Headers headers) {
```
use a package-private constructor if you call it from outer-class

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:
```diff
@@ -498,4 +477,46 @@ static long decodeULE128(ByteBuf in, long result) throws Http2Exception {
+            this.headers = headers;
+        }
+
+        private void addHeader(CharSequence name, CharSequence value) throws Http2Exception {
```
make package-private if called from outclass.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java:
```diff
@@ -22,64 +22,89 @@
             return value;
         }
 
+        public PseudoHeaderType getType() {
```
remove `get`  as we also not have it on value etc.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java:
```diff
@@ -22,64 +22,89 @@
+            return PSEUDO_HEADERS.get(header);
+        }
+
+        public static boolean hasPseudoHeaderFormat(CharSequence header) {
```
add java docs to the two methods.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:
```diff
@@ -498,4 +477,46 @@ static long decodeULE128(ByteBuf in, long result) throws Http2Exception {
         throw DECODE_ULE_128_DECOMPRESSION_EXCEPTION;
     }
+
+    private final class HeadersDecodingState {
```
static ?

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java:
```diff
@@ -152,4 +162,29 @@ protected final boolean validateHeaders() {
+                }
+
+                if (pseudoHeader.getType() == Http2Headers.PseudoHeaderType.REQUEST) {
+                    isRequest = true;
```
just replace this by:

```java
if (isResponse) {
    throw connectionError(PROTOCOL_ERROR, "Mix of request and response pseudo-headers.");
}
```

This way you can fast fail the loop.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java:
```diff
@@ -152,4 +162,29 @@ protected final boolean validateHeaders() {
+                if (pseudoHeader.getType() == Http2Headers.PseudoHeaderType.REQUEST) {
+                    isRequest = true;
+                } else {
+                    isResponse = true;
```
just replace this by:

```java
if (isRequest) {
    throw connectionError(PROTOCOL_ERROR, "Mix of request and response pseudo-headers.");
}
```

This way you can fast fail the loop.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java:
```diff
@@ -152,4 +162,29 @@ protected final boolean validateHeaders() {
+            }
+        }
+
+        if (isRequest && isResponse) {
```
This can be removed once you addressed the comments above

normanmaurer commented:
@madgnome thanks for the PR... I left a few comments.

Also please sign our ICLA:

https://netty.io/s/icla

madgnome commented:
ICLA signed.

fenik17 commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java:
```diff
@@ -152,4 +162,35 @@ protected final boolean validateHeaders() {
+                    throw connectionError(PROTOCOL_ERROR, "Invalid HTTP/2 pseudo-header '%s' encountered", key);
+                }
+
+                if (pseudoHeader.type() == Http2Headers.PseudoHeaderType.REQUEST) {
```
This can be simplified:
```
Http2Headers.PseudoHeaderType prevType = null;
for (...) {
   ...
   if (pseudoHeader.type() != prevType)) {
      requestAndResponsePseudoHeadersMixed();
   }
   prevType = pseudoHeader.type();
}

normanmaurer commented:
@madgnome I guess with this we can also re-enable the test here: https://github.com/netty/netty/blob/4.1/testsuite-http2/pom.xml#L76 ?

madgnome commented:
Indeed! will update.

normanmaurer commented:
@mosesn PTAL as well

mosesn commented:
Migrating to hard fail when connection headers are out of order makes me a bit nervous, but we can back it out and figure out a migration strategy if people try this in the wild and discover it's actually an issue.  🚢 🇮🇹

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Headers.java:
```diff
@@ -70,6 +70,7 @@ public void validateName(CharSequence name) {
     };
 
     private HeaderEntry<CharSequence, CharSequence> firstNonPseudo = head;
+    private final boolean keepInsertionOrder;
```
I would prefer to not have this in production code if it is only to support testing. I would prefer to isolate this to the unit tests if possible.

madgnome replied:
Yeah, wasn't sold on that either. 

Other solutions are:
* Duplicate DefaultHttp2Headers in test code
* Remove `final` on `DefaultHttp2Headers#newHeaderEntry` and extends it in test to implement an in insertion order iterator.

Any preference?

normanmaurer replied:
@madgnome I thin I would go with the "removing final".

Scottmitch replied:
Removing final sounds OK to me ... the question is how much would we have to duplicate in the tests? It sounds like what you want todo is different than what `DefaultHttp2Headers` is designed todo (enforce HTTP/2 constraints). `DefaultHttp2Headers` also doesn't provide much code besides the `Http2HeaderEntry` (which it looks like would be replaced anyways) and `HTTP2_NAME_VALIDATOR` (which is package private and maybe we can share?). So maybe duplication isn't too bad here?

madgnome replied:
I've changed the implementation of `DefaultHttp2HeadersDecoder` to instantiate `DefaultHttp2Headers` with `keepInsertionOrder` to `true`, doing that allows us to move all the validation to `DefaultHttp2HeadersDecoder` and to keep `HPackDecoder` untouched.

If you think `keepInsertionOrder` still shouldn't exist, I could implement a `DecoderHttp2Headers` class that would be use by the decoder to do that.

Scottmitch replied:
IIUC `keepInsertionOrder` is still only necessary for tests. If this is the case I would prefer to remove it and isolate this behavior to unit tests.

madgnome replied:
Oh damn, good catch. I was a bit too eager cleaning some code before commiting and remove the important part

```
protected Http2Headers newHeaders() {
        return new DefaultHttp2Headers(validateHeaders, (int) headerArraySizeAccumulator, true);
    }
```

`DefaultHttp2HeadersDecoder` will use `keepInsertionOrder`

Scottmitch replied:
ping me when behavior has been isolated to unit tests and I will take another look

madgnome replied:
You can see there https://github.com/netty/netty/pull/7500/files#diff-32bd1e577a26442a725a78acaf5822daR163 how we use `keepInsertionOrder` in prod code.

(Tests are back green)

madgnome replied:
@Scottmitch are you happy with `keepInsertionOrder` in `DefaultHttp2Headers` now that it's used by production code?

Otherwise I'll create an `DecoderHttp2Headers` extending `DefaultHttp2Headers` which keep insertion order and use that in the decoder.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java:
```diff
@@ -152,4 +162,24 @@ protected final boolean validateHeaders() {
     }
+
+    protected void validate(Http2Headers headers) throws Http2Exception {
+        Http2Headers.PseudoHeaderType prevType = null;
```
in addition to verifying the type is the same as the previous, shouldn't we also provide an "expected type" to satisfy the following requirements:

> Pseudo-header fields defined for requests MUST NOT appear in responses;
pseudo-header fields defined for responses MUST NOT appear in requests.

madgnome replied:
I'm going through h2spec failing tests and trying to fix them. The following tests should cover that

```
            <excludeSpec>8.1.2.3 - Sends a HEADERS frame with empty ":path" pseudo-header field</excludeSpec>
            <excludeSpec>8.1.2.3 - Sends a HEADERS frame that omits ":method" pseudo-header field</excludeSpec>
            <excludeSpec>8.1.2.3 - Sends a HEADERS frame that omits ":scheme" pseudo-header field</excludeSpec>
            <excludeSpec>8.1.2.3 - Sends a HEADERS frame that omits ":path" pseudo-header field</excludeSpec>
```

normanmaurer replied:
@madgnome @Scottmitch  +1 I think we can fix these step by step

Scottmitch replied:
I brought it up because it is mentioned in the commit message. If the intention is not to fix this in the current PR, please remove this text from the commit message.

madgnome replied:
We check that pseudo headers are not mixed (no request header with response header or vice versa ) so it does meet the requirement in a way

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:
```diff
@@ -121,13 +122,15 @@
         CharSequence name = null;
         IndexType indexType = IndexType.NONE;
+
+        HeadersDecodingState headersDecodingState = new HeadersDecodingState(streamId, headers,
```
can we avoid addition object allocation in the headers decode operation? This maybe a frequently called operation and if we can avoid intermediate allocations that would be preferred.

normanmaurer replied:
+1... would be nice if we could just store these directly

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:
```diff
@@ -498,4 +478,51 @@ static long decodeULE128(ByteBuf in, long result) throws Http2Exception {
+
+            boolean pseudoHeaderFormat = hasPseudoHeaderFormat(name);
+            if (sawRegularHeader && pseudoHeaderFormat) {
+                throw connectionError(PROTOCOL_ERROR, "Pseudo-header field found after regular header.");
```
It seems like we validate the pseudo header requirements in `DefaultHttp2HeadersDecoder` and in here in `HpackDecoder`. Can we validate in one or the other instead of both?

madgnome replied:
Agree, I wanted to put everything into `DefaultHttp2HeadersDecoder` but the way `DefaultHttp2Headers` works we lose the insertion order. 

Do you think it would be better to have everything in `HpackDecoder` or should I find a way to retrieve insertion order from DefaultHttp2Headers?

madgnome commented:
Ok, I've changed things a bit:

- I've undo all my changes to `HPackDecoder` 
- I've moved the validation for out of order headers to `DefaultHttp2HeadersDecoder` (now all the validation happen in the same place 🎆 )
-  For this to work it uses `keepInsertionOrder` in `DefaultHttp2Headers`. Another solution would be to create a `DecodedHttp2Headers` extending `DefaultHttp2Headers` which keep insertion order.

mosesn commented:
This approach seems reasonable to me.

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java:
```diff
@@ -150,6 +160,33 @@ protected final boolean validateHeaders() {
      */
     protected Http2Headers newHeaders() {
-        return new DefaultHttp2Headers(validateHeaders, (int) headerArraySizeAccumulator);
+        return new DefaultHttp2Headers(validateHeaders, (int) headerArraySizeAccumulator, true);
```
note that headers are modifiable by default. So after the headers are decoded they maybe modified and conceivably add another pseudo header (e.g. a proxy use case). I'm hesitant to generate a `Http2Headers` objects which does not enforce the iteration order required by the specification under these constraints. IMHO it would be safer just to isolate this behavior to the unit tests if that is where it is required.

normanmaurer replied:
@Scottmitch @madgnome +1

madgnome replied:
Oh damn, excellent point.

> IMHO it would be safer just to isolate this behavior to the unit tests if that is where it is required.

We need to know the insertion order to be able to detect if the headers are out of order when we decode them. 

Back to the drawing board...

madgnome commented:
Ok, I gave it another go.

Changes compared to previous commits:

- Validation is done entirely in `HPackDecoder`. I tried my best to limit object allocation
- New `InOrderHttp2Headers` class in test package to help with testing (No modification to existing `DefaultHttp2Headers` behaviour

normanmaurer commented:
@ejona86 @carl-mastrangelo any comment ?

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:
```diff
@@ -382,6 +396,28 @@ private void setDynamicTableSize(long dynamicTableSize) throws Http2Exception {
 
+    private Http2Headers.HeaderType validate(CharSequence name, Http2Headers.HeaderType previousHeaderType)
+        throws Http2Exception {
+        if (hasPseudoHeaderFormat(name)) {
```
Netty typically allows users to opt out of header validation which may otherwise add overhead. For example in a "trusted" scenario it may be desirable to avoid the additional overhead (map lookup, string inspection, etc...). I would suggest we propagate the `validateHeaders` parameter in `DefaultHttp2HeadersDecoder` to this class and conditionally do the validation.

madgnome replied:
- [X] Propage `validateHeaders`

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java:
```diff
@@ -35,51 +46,78 @@
             return value;
         }
 
+        public HeaderType type() {
```
does `HeaderType` need to be public? From the perspective of a `PseudoHeaderName` it is only allowed in either `request` or `response`, correct? If that is the case can we just have a method like this:

```java
// true -> the pseudo header can only be used in requests, false -> the pseudo header can only be used in responses
public boolean isRequestOnly() {
}
```

madgnome replied:
`HeaderType` can be `REGULAR_HEADER` , `REQUEST_PSEUDO_HEADER` and `RESPONSE_PSEUDO_HEADER`, we use it to track the current header type in the decoder to help with validation. 

We can make `HeaderType` package private or if we absolutely don't want it, we could try to rely on `Http2Headers.PseudoHeaderName` only but that would make the code more confusing I think.

madgnome replied:
Yeah makes perfect sense, I'll move `HeaderType` to `HPackDecoder` .

madgnome replied:
- [X] Move HeaderType to HPackDecoder

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/HpackDecoderTest.java:
```diff
@@ -512,4 +512,91 @@ public void testAccountForHeaderOverhead() throws Exception {
+            hpackEncoder.encodeHeaders(1, in, toEncode, NEVER_SENSITIVE);
+
+            // decode the headers, we should get an exception, but
+            // the decoded headers object should contain all of the headers
```
this makes it sound like you would expect `decoded` to contain `:test=1` after the decode operation, but that is not the case. Can you clarify this comment?

madgnome replied:
- [X] Remove comment

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/HpackDecoderTest.java:
```diff
@@ -512,4 +512,91 @@ public void testAccountForHeaderOverhead() throws Exception {
+            expectedException.expect(Http2Exception.class);
+            hpackDecoder.decode(1, in, decoded);
+        } finally {
+            in.release();
```
should we verify that `decoded` contains `:status=200`?

madgnome replied:
I think it's only necessary if we expect consumers to use the `Http2Headers` after a failed decoding attempt. It's currently not the case in `DefaultHttp2HeadersDecoder`

Scottmitch replied:
sgtm

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/HpackDecoderTest.java:
```diff
@@ -512,4 +512,91 @@ public void testAccountForHeaderOverhead() throws Exception {
+            expectedException.expect(Http2Exception.class);
+            hpackDecoder.decode(1, in, decoded);
+        } finally {
+            in.release();
```
should we verify that `decoded` contains `:method=200`?

carl-mastrangelo commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java:
```diff
@@ -35,51 +46,78 @@
+         * @return {@code true} if the header follow the pseudo-header format
+         */
+        public static boolean hasPseudoHeaderFormat(CharSequence header) {
+            return header.length() > 0 && header.charAt(0) == PSEUDO_HEADER_PREFIX;
```
Another idea: skip the header.length() check and just catch the exception on index out of bounds.  Most headers will not need the check.

madgnome replied:
- [X] Add optimisation for AsciiString

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java:
```diff
@@ -35,51 +46,86 @@
-        STATUS(":status");
+        STATUS(":status", RESPONSE_PSEUDO_HEADER);
+
+        private static char PSEUDO_HEADER_PREFIX = ':';
```
make this variable `final`.

madgnome replied:
- [x] Make final

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java:
```diff
@@ -35,51 +46,86 @@
+        public static boolean hasPseudoHeaderFormat(CharSequence header) {
+            try {
+                if (header instanceof AsciiString) {
+                    return ((AsciiString) header).byteAt(0) == PSEUDO_HEADER_PREFIX;
```
consider adding a new `private static final byte PSEUDO_HEADER_PREFIX_BYTE = (byte) ';';` and using it here to avoid any casting.

madgnome replied:
- [x] add byte format for prefix

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java:
```diff
@@ -35,51 +46,86 @@
+         *
+         * @return {@code true} if the header follow the pseudo-header format
+         */
+        public static boolean hasPseudoHeaderFormat(CharSequence header) {
```
nit: `header` -> `headerName`?

Scottmitch replied:
potential optimization: we should use this in [DefaultHttp2Headers](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Headers.java#L200) and use [HeadersBenchmark](https://github.com/netty/netty/blob/4.1/microbench/src/main/java/io/netty/microbench/headers/HeadersBenchmark.java#L51) to ensure there is benefit of this approach.

madgnome replied:
- [x] use this in `DefaultHttp2Headers`
- [x] benchmark

 I've modified the code a bit after benchmarking it

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:
```diff
@@ -398,14 +438,12 @@ private CharSequence readName(int index) throws Http2Exception {
         if (index - HpackStaticTable.length <= hpackDynamicTable.length()) {
-            HpackHeaderField hpackHeaderField = hpackDynamicTable.getEntry(index - HpackStaticTable.length);
-            return addHeader(headers, hpackHeaderField.name, hpackHeaderField.value, headersLength);
+            return  hpackDynamicTable.getEntry(index - HpackStaticTable.length);
```
nit: remove double space after `return`

madgnome replied:
- [X] remove space

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/HpackDecoderTest.java:
```diff
@@ -160,27 +164,13 @@ public void testDecodeULE128LongOverflow2() throws Http2Exception {
-        try {
-            final long expectedHeaderSize = 4026531870L; // based on the input above
-            hpackDecoder.setMaxHeaderTableSize(expectedHeaderSize);
-            hpackDecoder.decode(0, in, mockHeaders);
```
is there a reason for the method movement? if not, can you undo this movement of this unit test and just add the extra parameter here?

madgnome replied:
good catch, I'll move that back.

madgnome replied:
- [x] Move

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/HpackDecoderTest.java:
```diff
@@ -430,40 +454,6 @@ public void testLiteralNeverIndexedWithLargeValue() throws Http2Exception {
-            // the decoded headers object should contain all of the headers
-            Http2Headers decoded = new DefaultHttp2Headers();
-            try {
-                hpackDecoder.decode(1, in, decoded);
```
is there a reason for the method movement? if not, can you undo this movement of this unit test and just add the extra parameter here?

madgnome replied:
good catch, I'll move that back.

madgnome replied:
- [x] Move

Scottmitch commented on codec-http2/src/test/java/io/netty/handler/codec/http2/HpackDecoderTest.java:
```diff
@@ -520,7 +510,126 @@ public void testIncompleteHeaderFieldRepresentation() throws Http2Exception {
+            hpackEncoder.encodeHeaders(1, in, toEncode, NEVER_SENSITIVE);
+
+            // decode the headers, we should get an exception, but
+            // the decoded headers object should contain all of the headers
```
I still don't completely understand this comment. I guess it is trying to say `:status` should still be contained in `decoded` ... which is why I previously asked if we should validate the header contents after the fact. However if this is not necessary (which I agree it doesn't seem necessary) can we remove/clarify this part of the comment?

Same question in the tests below too.

madgnome replied:
Bad copy/paste, will delete

madgnome replied:
- [x] delete comment

Scottmitch commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java:
```diff
@@ -35,51 +35,84 @@
+            return PSEUDO_HEADERS.get(header);
+        }
+
+        public boolean isRequestOnly() {
```
can you plz add some javadocs here?

madgnome replied:
- [x] add javadoc

madgnome commented:
All feedbacks were actioned. 🤞

Scottmitch commented:
4.1 (3e6b54bb59)

Scottmitch commented:
thanks for your patience @madgnome !

normanmaurer commented:
👍

> Am 30.01.2018 um 04:43 schrieb Scott Mitchell <notifications@github.com>:
> 
> thanks for your patience @madgnome !
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
